src/main/kotlin/com/example/springrestuser/controllers/PublicAssetController.kt

```
@@ -0,0 +1,109 @@
+import com.example.springrestuser.services.UserAssetService
+import org.hibernate.exception.DataException
+import org.springframework.web.bind.annotation.*
+import java.lang.Exception
+import java.util.*
+
+@RestController
+class PublicAssetController(
+    private val service: PublicAssetService,
+) {
+
+    @PostMapping("/asset/public/")
+    fun add(@RequestBody asset: PublicAssetDTO): PublicAssetDTO {
+        if (validate(asset)) {
+            return toDTO(service.add(toDO(asset)))
+        } else {
+            throw RuntimeException("Invalid input data")
+        }
+    }
+
+    @GetMapping("/asset/public/{id}")
+    fun get(@PathVariable id: Int): PublicAssetDTO {
+        return toDTO(service.get(id))
+    }
+
+    @GetMapping("/asset/public/")
+    fun getAll(): List<PublicAssetDTO> {
+        val list = service.getAll()
+        val res = mutableListOf<PublicAssetDTO>()
+        for (asset in list) {
+            res.add(toDTO(asset))
+        }
+        return res
+    }
+
+    @PutMapping("/asset/public/{id}")
+    fun modify(@PathVariable id: Int, @RequestBody asset: PublicAssetDTO) {
+        if (validate(asset)) {
+            asset.id = id
+            service.modify(toDO(asset))
+        } else {
+            throw RuntimeException("Invalid input data")
+        }
+
+    }
+
+    @DeleteMapping("/asset/public/{assetId}")
+    fun remove(@PathVariable assetId: Int) {
+        service.delete(assetId)
+    }
+
+    fun validate(asset: PublicAssetDTO): Boolean {
+        var res = true
+        val providers = mutableListOf<String>()
+        for (provider in AssetSubtype.values()) {
+            providers.add(provider.toString())
+        }
+
+        if (asset.price < 0) {
+            res = false
+        }
+        if (asset.type.lowercase(Locale.getDefault()) != "public") {
+            res = false
+        }
+        if (asset.provider !in providers) {
+            res = false
+        }
+        return res
```

you could throw an exception here so that you won't need to introduce business logic in the controller